presto-main/src/main/java/com/facebook/presto/sql/planner/TranslationMap.java

```diff
@@ -131,14 +136,23 @@ public Symbol get(Expression expression)
     {
         if (expression instanceof FieldReference) {
             int field = ((FieldReference) expression).getFieldIndex();
-            Preconditions.checkArgument(fieldSymbols[field] != null, "No mapping for field: %s", field);
+            checkArgument(fieldSymbols[field] != null, "No mapping for field: %s", field);
             return fieldSymbols[field];
         }
 
         Expression translated = translateNamesToSymbols(expression);
+        checkArgument(expressionToSymbols.containsKey(translated), "No mapping for expression: %s", expression);
+        return expressionToSymbols.get(translated);
+    }
 
-        Preconditions.checkArgument(expressionMappings.containsKey(translated), "No mapping for expression: %s", expression);
-        return expressionMappings.get(translated);
+    public void put(Expression expression, Expression rewritten)
+    {
+        expressionToExpressions.put(translateNamesToSymbols(expression), rewritten);
+    }
+
+    public void setExpressionAsAlreadyTranslated(Expression node)
```

Comment:
I am going to add a javadoc. The motivation behind this method comes from corner case sql where  we have the same subquery expression in two different in predicate expressions. See the example:

```
SELECT * FROM x 
WHERE x.col1 IN (SELECT y.col1 FROM y) 
  AND x.col2 IN (SELECT y.col1 FROM y)
```

Take a look what happens in regards of translation map when analyze `WHERE` expression. A new translation `subquery_1` is added for first `(SELECT y.col1 FROM y)`, then a translation `subquery_2` is added for second `(SELECT y.col1 FROM y)` which overwrites the first. In this case the query is rewritten to something like:

```
SELECT * FROM x 
WHERE x.col1 IN subquery_2
  AND x.col2 IN subquery_2

where: 
subquery_2: SELECT y.col1 FROM y
```

In that case I am not able to properly rewrite `Apply` nodes as I have two Apply nodes but only one matching IN predicate symbols (x.col2, subquery_2). The other one matches to (x.col1, subquery_1).

I solved this by following approach. I manually create an IN predicate expression by creating subquery symbol. Then I set in translation map that this (representing subquery) symbol is already translated. This prevents the translation map from looking for a filed for this symbol reference in Analysis etc. TranslationMap tries to rewrite all `QualfiedNameReference` objects when it is unable to do so, it fails.

As I write this I thought I can try with `IdentityHashMap` for storing translation it could work as well. However this could introduce higher risk that something get broken somewhere else in the code..